chain/chunks/src/lib.rs

```diff
@@ -328,31 +362,23 @@ impl ShardsManager {
                 ..
             } = chunk_header;
             let chunk_hash = chunk_header.chunk_hash();
+            self.request_chunk(parent_hash, chunk_hash, shard_id, height, chunk_header)?;
+        }
+        Ok(())
+    }
 
-            if self.requested_partial_encoded_chunks.contains_key(&chunk_hash) {
-                continue;
-            }
-
-            self.encoded_chunks.get_or_insert_from_header(chunk_hash.clone(), Some(&chunk_header));
-
-            self.requested_partial_encoded_chunks.insert(
-                chunk_hash.clone(),
-                ChunkRequestInfo {
-                    height,
-                    parent_hash,
-                    shard_id,
-                    last_requested: Instant::now(),
-                    added: Instant::now(),
-                },
-            );
-            self.request_partial_encoded_chunk(
-                height,
-                &parent_hash,
-                shard_id,
-                &chunk_hash,
-                false,
-                false,
-            )?;
+    pub fn request_chunks_for_orphan(
+        &mut self,
+        parent_hash: CryptoHash,
+        chunks_to_request: Vec<ShardChunkHeader>,
+    ) -> Result<(), Error> {
+        for chunk_header in chunks_to_request {
+            let ShardChunkHeader {
+                inner: ShardChunkHeaderInner { shard_id, height_created: height, .. },
+                ..
+            } = chunk_header;
+            let chunk_hash = chunk_header.chunk_hash();
+            self.request_chunk(parent_hash, chunk_hash, shard_id, height, chunk_header)?;
```

Comment:
Here I would add a check that `epoch(parent_hash) == epoch(chunk)` and bail out otherwise, because if we initiate a request with a wrong epoch id, we might end up never retrieving the chunks.

Reply:
I don't think it's possible to get epoch_id from chunk.prev_block_hash.
We can calculate epoch_id only from blocks which are already accepted. `runtime_adapter.get_epoch_id_from_prev_block` relies on it completely. (It's a BlockInfo / get_epoch_and_valset issue, we discussed offline a lot.)

Reply:
disregard this, I can take epoch_id from orphan block